### PR TITLE
Prebid Analytics API

### PR DESCRIPTION
**Pull request by protonate**
Provides two new Prebid API methods to register analytics adapters and enable analytics tracking for any analytics module.
* `pbjs.enableAnalytics()` is called once on the page after all analytics libraries have been loaded
* `pbjs.registerAnalyticsAdapter` is available to register an adapter dynamically (optional)

An analytics provider can be configured with these steps:

1. Add the analytics library to your page as described in the provider's documentation (see `integrationExamples/gpt/examples...` for Google Analytics.)
1. Add a call to `pbjs.enableAnalytics(analyticsAdapters)`:
```
pbjs.que.push(function () {
      pbjs.enableAnalytics([{
          provider: 'ga',
          options: {
              enableDistribution: false
          }
      }, {
              provider: 'other',
              options: {
                      foo: 1234
              }
      }
      ]);
  });
```
1. Create an analytics adapter to listen for Prebid events and call the analytics library (see `src/adapters/analytics/example.js`)
1. Add the analytics adapter file name to the "analytics" array in `package.json`.

updated following code review with Product
updated 4/24 with factory function pattern

### CODE REVIEW

**Comment by protonate** [comment]
@mkendall07 This is ready to discuss, I still have to address calling `pbjs.enableAnalytics` twice on the page load, let me know what you think of the approach.

**Comment by protonate** [comment]
@mkendall07 I've updated to register analytics adapters via entries in `package.json` either with an `options` for configuration or as just the adapter file name.

This uses a technique in `analyticsLoader.js` to write data to the stream using JSON that could be useful in other cases as well. See [here](https://github.com/prebid/Prebid.js/pull/297/files#diff-36a1ef90bf7c133fb3962b26886cd623R46).

Ready for review.

File: integrationExamples/gpt/gpt_aliasingBidder.html [hunk 1]
```diff
@@ -2,141 +2,160 @@
...
-            console.log('Targeting params:')
-            console.log(pbjs.getAdserverTargeting());
+    <script>
+        googletag.cmd.push(function () {
+
+            var rightSlot = googletag.defineSlot('/9968336/header-bid-tag-0', [
+                [
+                    300,
+                    250
+                ],
```
Comment (mkendall07) [inline integrationExamples/gpt/gpt_aliasingBidder.html]: these look really ugly, can you put them back into 1 line?

**Comment by protonate** [comment]
Ok formatting rolled back.

**Comment by protonate** [comment]
@mkendall07 @Heray updated following discussion, rebased and ready for review.

**Comment by mkendall07** [comment]
thanks @protonate We will hold off on merge until after release.

File: loaders/analyticsLoader.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                || require('./adapters/analytics/${adapter}.js');
+        exports.registerAnalyticsAdapter({ adapter: ${adapter}, code: '${adapter}' });\n`;
+    }).join('');
+  }
+};
+
+module.exports = blockLoader(options);
+
+// In progress: a means to write loader code for libraries
+function loadLibrary(adapter) {
```
Comment (mkendall07) [inline loaders/analyticsLoader.js]: @protonate 
What's the advantage of doing this build time vs load time via `loadScript` etc?

File: src/adapters/analytics/example2.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+export default utils.extend(adapter(
+  {
+    global,
+    url,
+    handler
+  }
+),
+  {
+  track({ type, data }) {
```
Comment (mkendall07) [inline src/adapters/analytics/example2.js]: Can you add some comments about how this works for overriding default functionality?

File: src/adapters/analytics/ga.js [hunk 4]
```diff
@@ -23,21 +23,17 @@ var _trackerSend = null;
...
  * @return {[type]}    [description]
  */
-exports.enableAnalytics = function (gaOptions) {
-  if (typeof gaOptions.global !== 'undefined') {
-    _gaGlobal = gaOptions.global;
-  } else {
-    //default global is window.ga
-    _gaGlobal = 'ga';
-  }
+exports.enable = function ({ provider, options }) {
```
Comment (mkendall07) [inline src/adapters/analytics/ga.js]: Was this suppose to be renamed to `enable`?

File: src/adaptermanager.js [hunk 5]
```diff
@@ -82,7 +85,41 @@ exports.aliasBidAdapter = function (bidderCode, alias) {
...
+    utils.logError('Prebid Error: analyticsAdapter or analyticsCode not specified');
+  }
+};
+
+exports.enableAnalytics = function (config) {
+  if (!utils.isArray(config)) {
+    config = [config];
+  }
+
+  utils._each(config, adapterConfig => {
```
Comment (mkendall07) [inline src/adaptermanager.js]: We should probably offer some protection against calling `adapter.enable` multiple times (so not to duplicate data sent)

**Comment by protonate** [comment]
Closed in favor of #349

**Comment by protonate** [comment]
Reopened now that PR #349 has merged.

**Comment by protonate** [comment]
@mkendall07 this is rebased onto master and ready to review.

**Comment by protonate** [comment]
@matthewlane @mkendall07 this is updated and can be reviewed, I'll push tests up shortly.

File: src/adaptermanager.js (hunk 5 above)
Comment (protonate) [inline src/adaptermanager.js]: I added code to replace `enableAnalytics` with a warning here:
https://github.com/prebid/Prebid.js/pull/297/files#diff-84954ceef9a26dfb9a93dfafbd10c8e4R69

**Comment by protonate** [comment]
@mkendall07 review notes addressed

**Comment by mkendall07** [comment]
@protonate 
Overall I think this looks good. Is there a pattern for inheriting from `AnalyticsAdapter.js` but not using a `url` to load the library and all the "brains" are in the adapter itself? I think that might be a useful example as well (and most likely path for AN adapter). 

For `AnalyticsAdapter.js` it would be useful for some basic methods to send POST/GET request easily to any endpoint.

File: src/adapters/analytics/AnalyticsAdapter.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+  function _enqueue({ eventType, args }) {
+    const _this = this;
+
+    if (global && window[global] && eventType && args) {
+      _track({ eventType, args });
+    } else {
+      _queue.push(function () {
+        _eventCount++;
+        _this.track({ eventType, args });
```
Comment (mkendall07) [inline src/adapters/analytics/AnalyticsAdapter.js]: There seems to be a problem capturing `this` properly. When it's invoked via the script load callback, it's undefined.
image

**Comment by protonate** [comment]
@mkendall07 this has been updated to address the `this` _undefined_ and to add support for an analytics adapter using an endpoint instead of a library.

**Comment by protonate** [comment]
@mkendall07 rebased onto master

File: gulpfile.js [hunk 7]
```diff
@@ -25,6 +25,24 @@ var prebid = require('./package.json');
 var dateString = 'Updated : ' + (new Date()).toISOString().substring(0, 10);
 var packageNameVersion = prebid.name + '_' + prebid.version;
 var banner = '/* <%= prebid.name %> v<%= prebid.version %>\n' + dateString + ' */\n';
+var analyticsDirectory = '../analytics';
+
+var _options = // use for production, more info on options https://davidwalsh.name/compress-uglify
```
Comment (mkendall07) [inline gulpfile.js]: remove `_options`

File: gulpfile.js [hunk 8]
```diff
@@ -58,10 +76,10 @@ gulp.task('webpack', function () {
 
   webpackConfig.devtool = null;
 
-  return gulp.src(['src/prebid.js'])
+  return gulp.src([...helpers.getAnalyticsSources(analyticsDirectory), 'src/prebid.js'])
     .pipe(webpack(webpackConfig))
     .pipe(replace('$prebid.version$', prebid.version))
-    .pipe(uglify())
+    .pipe(uglify(_options))
```
Comment (mkendall07) [inline gulpfile.js]: drop `_options` here

File: package.json [hunk 9]
```diff
@@ -119,5 +122,7 @@
     "webpack-stream": "^3.1.0",
     "yargs": "^1.3.1"
   },
-  "dependencies": {}
+  "dependencies": {
```
Comment (mkendall07) [inline package.json]: try to remove

File: package.json [hunk 10]
```diff
@@ -43,6 +43,9 @@
       }
     }
   ],
+  "analytics": [
+    "appnexus"
```
Comment (mkendall07) [inline package.json]: empty array.

**Comment by mkendall07** [comment]
Thanks @matthewlane 
Good to merge.

**Pull request merged**